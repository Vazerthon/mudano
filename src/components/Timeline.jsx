import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { format, isEqual } from 'date-fns';

import Username from './Username';
import DayCell, { entryPropType } from './DayCell';

const Container = styled.div`
  display: flex;
  height: 63vh;
  align-items: flex-end;
`;

const Scroller = styled.div`
  overflow-x: auto;
  display: flex;
  align-items: flex-end;
  height: 100%;
`;

const DateTimeline = styled.div`
  display: flex;
`;

const DateColumn = styled.div`
  display: flex;
  flex-direction: column;
`;

const DateLabel = styled.span`
  transform: rotate(-60deg) translateX(50px) translateY(-77px);
  width: 200px;
  margin-right: -180px;
`;

const Usernames = styled.div`
  display: flex;
  flex-direction: column;
  margin-bottom: 16px;
`;

const HighlightOverlay = styled.div`
  position: fixed;
  background-color: green;
  height: ${({ height }) => `${height}px`};
  width: 20px;
  margin-top: 18px;

  opacity: 0;

  :hover {
    opacity: 0.4;
  }
`;

const matchingEntries = (entries, date) =>
  entries.filter(e => isEqual(e.date, date));

const prettyDate = date => format(date, 'Do MMM YY');

function Timeline({ users, user, timeline }) {
  return (
    <Container>
      <Usernames>
        {users.map(u => (
          <Username key={u.userId} highlight={u.userId === user.userId}>
            {u.name}
          </Username>
        ))}
      </Usernames>
      <Scroller>
        <DateTimeline>
          {timeline.map(d => (
            <DateColumn key={d.date}>
              <HighlightOverlay height={(users.length + 1) * 20} />
              <DateLabel>{prettyDate(d.date)}</DateLabel>
              {users.map(u => (
                <DayCell
                  key={`${u.userId}-${prettyDate(d.date)}`}
                  entries={matchingEntries(u.entries, d.date)}
                  isPublicHoliday={d.isPublicHoliday}
                />
              ))}
            </DateColumn>
          ))}
        </DateTimeline>
      </Scroller>
    </Container>
  );
}

export const userPropType = PropTypes.shape({
  userId: PropTypes.string,
  name: PropTypes.string,
  entries: PropTypes.arrayOf(entryPropType),
});

export const usersPropType = PropTypes.arrayOf(userPropType);
export const timelinePropType = PropTypes.arrayOf(
  PropTypes.shape({ date: PropTypes.date, isPublicHoliday: PropTypes.bool }),
);

Timeline.propTypes = {
  users: usersPropType,
  user: userPropType,
  timeline: timelinePropType,
};

Timeline.defaultProps = {
  users: [],
  user: {},
  timeline: [],
};

export default Timeline;
