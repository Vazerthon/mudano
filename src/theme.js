const theme = {
  colour: {
    primary: '#1e252b',
    secondary: '#dcd6ca',
    highlight: '#9a3cbb',
    timeline: {
      P: '#0971b2', // public holiday
      V: '#54cc14', // vacation
      T: '#cc4a14', // training
      present: '#9a3cbb',
    },
  },
  layout: {
    padding: '8px',
  },
};

export default theme;
