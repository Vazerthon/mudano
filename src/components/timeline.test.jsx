/* eslint-env jest */

import React from 'react';
import shallowThemeWrapper from './test-helper';
import theme from '../theme';

import Timeline from './timeline';

const shallow = component => shallowThemeWrapper(component, theme);

describe('timeline component', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<Timeline />);
    expect(wrapper).toMatchSnapshot();
  });
});
