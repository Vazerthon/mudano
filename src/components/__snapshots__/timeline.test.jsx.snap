// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`timeline component renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Timeline
    timeline={Array []}
    users={Array []}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <styled.div>
          
</styled.div>,
        <styled.div>
          <styled.div>
                    
          </styled.div>
</styled.div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [],
        },
        "ref": null,
        "rendered": Array [],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <styled.div>
            
</styled.div>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [],
          },
          "ref": null,
          "rendered": Array [],
          "type": [Function],
        },
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <styled.div>
            
</styled.div>,
          <styled.div>
            <styled.div>
                        
            </styled.div>
</styled.div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [],
          },
          "ref": null,
          "rendered": Array [],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <styled.div>
              
</styled.div>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [],
            },
            "ref": null,
            "rendered": Array [],
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "context": Object {
      "__styled-components__": [Function],
      "__styled-components__next__": Object {
        "getTheme": [Function],
        "subscribe": [Function],
        "unsubscribe": [Function],
      },
    },
  },
}
`;
